src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+    fn try_add_pending_request(
+        &self,
+        reqid: UUID,
+    ) -> Result<Receiver<UMessage>, ServiceInvocationError> {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Entry::Vacant(entry) = pending_requests.entry(reqid) {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                entry.insert(tx);
+                Ok(rx)
+            } else {
+                Err(ServiceInvocationError::AlreadyExists(
+                    "RPC request with given ID already pending".to_string(),
+                ))
+            }
+        } else {
+            Err(ServiceInvocationError::RpcError(UStatus::fail_with_code(
+                UCode::INTERNAL,
+                "failed to add response handler",
+            )))
+        }
+    }
+
+    fn handle_response(&self, reqid: &UUID, response_message: UMessage) {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Some(sender) = pending_requests.remove(reqid) {
+                let _ = sender.send(response_message);
+            } else {
+                // we seem to have received a duplicate of the response message,
+                // ignoring it
+            }
+        }
+    }
+
+    fn remove_pending_request(&self, reqid: &UUID) -> Option<Sender<UMessage>> {
+        self.pending_requests
+            .lock()
+            .map_or(None, |mut pending_requests| pending_requests.remove(reqid))
+    }
+
+    #[cfg(test)]
+    fn contains(&self, reqid: &UUID) -> bool {
+        self.pending_requests
+            .lock()
+            .map_or(false, |pending_requests| {
+                pending_requests.contains_key(reqid)
+            })
+    }
+}
+
+#[async_trait]
+impl UListener for ResponseListener {
+    async fn on_receive(&self, msg: UMessage) {
+        if msg.is_response() {
+            if let Some(reqid) = msg
+                .attributes
+                .as_ref()
+                .and_then(|attribs| attribs.reqid.clone().into_option())
+            {
+                self.handle_response(&reqid, msg);
+            }
+        }
+    }
+}
+
+/// An ['RpcClient'] which keeps all information about pending requests in memory.
+pub struct InMemoryRpcClient {
+    transport: Arc<dyn UTransport>,
+    uri_provider: Arc<dyn LocalUriProvider>,
+    response_listener: Arc<ResponseListener>,
+}
+
+impl InMemoryRpcClient {
+    /// Creates a new RPC client for a given transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `transport` - The uProtocol Transport Layer implementation to use for invoking service operations.
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the generic RPC Response listener could not be
+    /// registered with the given transport.
+    pub async fn new(
+        transport: Arc<dyn UTransport>,
+        uri_provider: Arc<dyn LocalUriProvider>,
```

Comment:
IIRC we left it open that e.g. UPClientZenoh could implement LocalUriProvider or impl this on a separate struct, right?

So in the former case this would look like:

```rust
let rpc_client = InMemoryRpcClient::new(up_client_zenoh.clone(), up_client_zenoh.clone());
```

right?

Are you thinking to add some recommendation on which path is better in the future as we learn more?

Reply:
This is correct. I am currently not in a position to recommend one or the other, so I'd rather let users decide what they prefer/see fit.